src/lib/scheduler/worker.cpp

```diff
@@ -128,17 +140,66 @@ void Worker::join() {
 uint64_t Worker::num_finished_tasks() const { return _num_finished_tasks; }
 
 void Worker::_wait_for_tasks(const std::vector<std::shared_ptr<AbstractTask>>& tasks) {
-  auto tasks_completed = [&tasks]() {
-    // Reversely iterate through the list of tasks, because unfinished tasks are likely at the end of the list.
-    for (auto it = tasks.rbegin(); it != tasks.rend(); ++it) {
-      if (!(*it)->is_done()) {
+  // This lambda checks if all tasks from the vector (our "own" tasks) have been executed. If they are, it causes
+  // _wait_for_tasks to return. If there are remaining tasks, it primarily tries to execute these. If they cannot be
+  // executed, the worker performs work for others (i.e., executes tasks from the queue).
+  auto all_own_tasks_done = [&]() {
+    auto all_done = true;
+    // Note: If we found a task that has not yet been executed, we reset this loop and start from the beginning.
+    // As such, both all_done and it may be reset outside of the following line.
+    auto it = tasks.begin();
+    while (it != tasks.end()) {
+      const auto& task = *it;
+      if (task->is_done()) {
+        ++it;
+        continue;
+      }
+
+      if (!task->is_ready()) {
+        // Task is not yet done - check if it is ready for execution
+        all_done = false;
+
+        // Task cannot be executed. We could stop here and simply return all_own_tasks_done == false. Instead, we
+        // continue in the list of tasks and check if there are any other tasks that we could work on.
+        ++it;
+        continue;
+      }
+
+      // Give other tasks, i.e., tasks that are unrelated to what we are currently waiting on, a certain chance of being
+      // executed, too. Anectotal evidence says that this is a good idea. For some reason, this keeps the memory
+      // consumption of TPC-H Q6 low even if the scheduler is overcommitted. Because generating random numbers is
```

Comment:
What consumption?

Reply:
Please clarify - what information beyond "memory consumption" would you need?

Reply:
Never mind. On a mobile "consumption" was not shown without scrolling.